text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1466 @@
+interface WireDecorator extends Decorator {
+    type: "wire";
+    adapterId: string;
+    // module name in canonical form
+    adapterModule?: string;
+    adapterConfig?: object;
+}
+
+// Information about a static import statement in a module.
+interface Import {
+    importType?: ('DefaultBinding' | 'NamedImports' | 'NamespacedImport')[];
+    importsList?: string[] | string;
+    // module name in canonical form
+    moduleSpecifier: string;
+    // Location starts from the import key word to the end of statement 
+    location: SourceLocation;
+}
+
+// Information about a dynamic import statement in a module.
+interface DynamicImport {
+    // module name in canonical form
+    moduleSpecifier: string;
+    moduleNameType: 'string' | 'unresolved';
+    location: SourceLocation;
+    hints?: DynamicImportHint[];
+}
+interface DynamicImportHint {
+    // Hint value with leading and trailing spaces of hint phrase trimmed
+    rawValue: string;
+    key: string;
+    value: string;
+    location: SourceLocation;
+}
+
+// Information about a export statement in a module.
+interface Export {
+    exportsList?: {
+        type: 'class' | 'function' | 'expression';
+        name?: 'string';
+    }[];
+    default?: boolean;
+    location: SourceLocation;
+}
+
+// Information about an export statement used to reexport bindings imported from another module.
+interface ReExport {
+    exportsList: {
+        name: string;
+        type: 'class' | 'function' | 'expression';
+    }[];
+    // module name in canonical form
+    moduleSpecifier: string;
+    location: SourceLocation;
+}
+
+// Information about a dom event instantiated in script.
+interface DOMEvent {
+    eventType: string;
+    isCustomEvent?: boolean;
+    options?: {
+        bubbles?: boolean;
+        composed?: boolean;
+    };
+    location: SourceLocation;
+}
+
+// Information about an event handled declaratively in the template.
+interface ProgrammaticEventListener {
+    type: string;
+    targetType: 'host' | 'shadowRoot' | 'Node';
+    options?: {
+        capture?: boolean;
+    };
+    location: SourceLocation;
+}
+
+// A parent component class including the 'LightningElement' class provided by the lwc module.
+type ParentClass =
+    | string // Another class declared in the same file
+    | Mixin
+    // or a class imported from an external module
+    | {
+          name: string;
+          // module name in full canonical form, the ModuleReference object can be looked up at 
+          // the root ScriptFile object using this name
+          moduleName?: string;
+          // Location of the parent class identifier
+          location: SourceLocation;
+      };
+
+// Information about a module imported into a module using the 'import' statement.
+interface ModuleReference {
+    // Full canonical name, eg. lightning/input, lightning/recordForm
+    name: string;
+    namespace?: string;
+    moduleIdentifier?:
```

Comment:
the notion of scoped modules is salesforce specific, i suggest that we don't introduce special treatment at the root level and treat every reference as a module. However, when we do encounter salesforce specific identifier, we can capture that information under a separate property 
```js
interface ModuleReference {
    // Full canonical name, eg. lightning/input, lightning/recordForm
    name: string;
    namespace?: string;
    type: 'module' | '@salesforce',
    sfdcResource?: {
         scoped: 'apexClass' | 'client' | 'slds' ... (the rest);
         namespacedId: string; // scoped module namespace normalization 
    }

```

Reply:
Can you elaborate how `namespace` and `namespacedId` different? I will incorporate that information into the RFC.